x/transferinject/keeper/keeper.go

```diff
@@ -0,0 +1,168 @@
+package keeper
+
+import (
+	"errors"
+	"fmt"
+	. "slices"
+
+	errorsmod "cosmossdk.io/errors"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	errortypes "github.com/cosmos/cosmos-sdk/types/errors"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+	transfertypes "github.com/cosmos/ibc-go/v6/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v6/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v6/modules/core/05-port/types"
+
+	rtypes "github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+	"github.com/dymensionxyz/dymension/v3/x/transferinject/types"
+)
+
+type Keeper struct {
+	cdc codec.BinaryCodec
+	porttypes.IBCModule
+	porttypes.ICS4Wrapper
+	rollappKeeper types.RollappKeeper
+	bankKeeper    types.BankKeeper
+}
+
+func NewTransferInject(
+	cdc codec.BinaryCodec,
+	rollappKeeper types.RollappKeeper,
+	bankKeeper types.BankKeeper,
+) *Keeper {
+	return &Keeper{
+		cdc:           cdc,
+		rollappKeeper: rollappKeeper,
+		bankKeeper:    bankKeeper,
+	}
+}
+
+func (t *Keeper) SetMiddleware(
+	ibc porttypes.IBCModule,
+	ics porttypes.ICS4Wrapper,
+) {
+	t.IBCModule = ibc
+	t.ICS4Wrapper = ics
+}
+
+// SendPacket wraps IBC ChannelKeeper's SendPacket function
+func (t *Keeper) SendPacket(
+	ctx sdk.Context,
+	chanCap *capabilitytypes.Capability,
+	destinationPort string, destinationChannel string,
+	timeoutHeight clienttypes.Height,
+	timeoutTimestamp uint64,
+	data []byte,
+) (sequence uint64, err error) {
+	rollapp, err := t.rollappKeeper.ExtractRollappFromChannel(ctx, destinationPort, destinationChannel)
+	if err != nil {
+		return 0, fmt.Errorf("cannot extract rollapp id from packet: %w", err)
+	}
+	// TODO: consider that other chains also might want this feature
+	if rollapp == nil {
+		return t.ICS4Wrapper.SendPacket(ctx, chanCap, destinationPort, destinationChannel, timeoutHeight, timeoutTimestamp, data)
+	}
+
+	packet := new(transfertypes.FungibleTokenPacketData)
+	if err = types.ModuleCdc.UnmarshalJSON(data, packet); err != nil {
+		return 0, errorsmod.Wrapf(errortypes.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet data: %s", err.Error())
+	}
+
+	// check if the rollapp already contains the denom metadata
+	if ContainsFunc(rollapp.TokenMetadata, func(dm *rtypes.TokenMetadata) bool { return dm.Base == packet.Denom }) {
```

Comment:
please DRY out this function with the acknowledgment one and document/explain it